### PR TITLE
[NVIDIA] New collection for variables 'overwrite_with_gradient'

### PR DESCRIPTION
**Pull request by kaixih**
This PR supports the new var collection name: 'overwrite_with_gradient'. Such variables will be store in `params` but their gradients won't go to the optimizers and won't have corresponding opt states. Instead, their gradients will be used as the new variables in the next step.

There are four related PRs, and should be reviewed in this order:
(1) https://github.com/google/praxis/pull/29
current-->(2) https://github.com/google/paxml/pull/48
(3) https://github.com/google/praxis/pull/28
(4) https://github.com/google/paxml/pull/49

cc. @pjannaty @reedwm @nluehr @lukaszlew

### CODE REVIEW

File: paxml/tasks_lib.py [hunk 1]
```diff
@@ -792,10 +792,19 @@ class CheckpointLoadingRules(NamedTuple):
...
+  """Returns whether each var is in over_with_gradient collection."""
+  return jax.tree_util.tree_map(
+      lambda x: 'overwrite_with_gradient' in x.collections, var_weight_hparams)
+
+
 def get_excluded_var_mask_for_grad_or_opt(
     var_weight_hparams: NestedJTensor,
     learner: learners_lib.Learner,
     mask_all_non_trainable: bool,
+    mask_all_overwrite: bool = False,
```
Comment (lukaszlew) [inline paxml/tasks_lib.py:807]: mask_all_overwrite_with_gradient

File: paxml/tasks_lib.py [hunk 2]
```diff
@@ -811,6 +820,14 @@ def get_excluded_var_mask_for_grad_or_opt(
     excluded_for_grad = py_utils.match_variable_names(
         var_weight_hparams, learner.bprop_variable_exclusion
     )
+  if mask_all_overwrite:
+    overwrite_var_mask = \
+        get_overwrite_with_gradient_var_mask(var_weight_hparams)
+    excluded_for_grad = jax.tree_map(
+        lambda is_overwrite, old: is_overwrite or old,
```
Comment (lukaszlew) [inline paxml/tasks_lib.py:827]: let's use better variable naming. Other functions are using 'e' for elements of excluded_for_grad

Comment (lukaszlew) [reply paxml/tasks_lib.py:827]: Let's follow the pattern and define base_layer.var_overwrite_with_gradient(x) function.
and pass var_weight_hparams to this tree_map directly, just like the tree_map below.

File: paxml/trainer_lib.py [hunk 3]
```diff
@@ -1142,6 +1149,13 @@ def train_step_single_learner(
         mdl_vars,
         vars_with_opt,
     )
+    # Update 'overwrite_with_gradient' params.
+    mdl_vars = jax.tree_map(
+        lambda e, new, old: new if e else old,
```
Comment (lukaszlew) [inline paxml/trainer_lib.py:1154]: can we pass hparams here and use the function (see other comment)

File: paxml/trainer_lib.py [hunk 4]
```diff
@@ -1130,8 +1130,15 @@ def train_step_single_learner(
     wps_with_opt = tasks_lib.filter_vars_for_grad_or_opt(
         var_weight_hparams, excluded_for_opt
     )
+    # Exclude the overwrite_with_gradient params, since they don't go through
+    # the optimizer and they don't have corresponding opt states.
+    overwrite_var_mask = \
+        tasks_lib.get_overwrite_with_gradient_var_mask(var_weight_hparams)
+    normal_grads = tasks_lib.filter_vars_for_grad_or_opt(
```
Comment (lukaszlew) [inline paxml/trainer_lib.py:1137]: the learner handels non-trainable variables by running this code

    var_is_learnable = jax.tree_util.tree_map(
        lambda x: not base_layer.var_not_trainable(x), var_weight_hparams
    )

We should either add another condition to the learner, or require all `overwrite_with_gradient` variables to be non-trainable. The latter makes sense to me. How about that?

**Review by lukaszlew — CHANGES_REQUESTED** [review]

File: paxml/trainer_lib.py (hunk 4 above)
Comment (kaixih) [reply paxml/trainer_lib.py:1137]: But do we need to pass the grads of 'overwrite_with_gradient' as part of `transformed_grads` and pass it to the apply_gradients(). So, something like:
```python
    transformed_grads, new_opt_states = learner.update_states(
        normal_grads, states.opt_states[0], vars_with_opt, wps_with_opt
    )
    transformed_grads = treemap(lambda a, b, c: b if a else c, overwrite_mask, transformed_grads, grads) 
    vars_with_opt = learner.apply_gradient(
        vars_with_opt, transformed_grads, wps_with_opt
    )
```
And inside the learners, I can add a branch to update the grads as you mentioned.

But I feel the above logic is more complicated, since we need to hijack the transformed_grads.

File: paxml/tasks_lib.py (hunk 2 above)
Comment (kaixih) [reply paxml/tasks_lib.py:827]: Since we still want the old way to provide the var_overwrite_with_gradient from base_layer from praxis, it seems the PRs from paxml and praxis still depend on each other. I will merge the changes and still do two PRs.

File: paxml/trainer_lib.py (hunk 4 above)
Comment (kaixih) [reply paxml/trainer_lib.py:1137]: The code you pasted is from the `apply_gradient()` of `learners.py` in paxml. But the `apply_gradient()` we used in [here](https://github.com/google/paxml/blob/main/paxml/trainer_lib.py#L1136) is from optax, since it doesn't accept the `var_weight_hparams`. So, unless we want to introduce `var_weight_hparams` to the apply_gradient() to optax optimizers, I don't think it is easy to put the logic of updating `overwrite_with_gradient` var inside learner.

Comment (kaixih) [reply paxml/trainer_lib.py:1137]: The above is wrong and `apply_gradient()` is indeed in the learners.py rather in optax.

However, I feel if we want to do the variable update inside the `apply_gradient()`, we would need to do lots of changes. Currently, the logic is like:
```
var_with_opt = ...
wps_with_opt = ...
grads_with_opt = ... # we added this
transformed_grads, new_opt_states = learner.update_states(grads_with_opt, states, vars_with_opt, wps_with_opt)
vars_with_opt = learner.apply_gradient(vars_with_opt, transformed_grads, wps_with_opt)
mdl_vars = vars_with_opt if has_grad else mdl_vars
mdl_vars = grads if is_overwrite else mdl_vars # we added this
```
Basically, we need to exclude the overwrite-with-grad params and at the very end, do the overwrite.

However, if we move the logic into the learner.apply_gradient(), the callsite of the above will be changed to:
```
var_with_opt = ...
wps_with_opt = ...
grads_with_opt = ... # we still add this
transformed_grads, new_opt_states = learner.update_states(grads_with_opt, states, vars_with_opt, wps_with_opt)
var_with_opt_and_overwrite = ... # we need to add this
transformed_grads = grads if is_overwrite else transformed_grads # we need to merge the grads
wps_with_opt_and_overwrite = ... # we need to add this
vars_with_opt = learner.apply_gradient(var_with_opt_and_overwrite, transformed_grads, wps_with_opt_and_overwrite)
mdl_vars = vars_with_opt if has_grad else mdl_vars
# mdl_vars = grads if is_overwrite else mdl_vars # Yes, we don't need this, however, we need to do more in apply_gradient()
```
Then, in the apply_gradient() we do:
```
exclude the overwrite from var_with_opt_and_overwrite, transformed_grads, wps_with_opt_and_overwrite # we need to add this
do the original updating
get the masks of overwrite for var_with_opt_and_overwrite, transformed_grads, wps_with_opt_and_overwrite # we need to add this
do the update for the overwrite # we need to add this
```
So, to summarize, I just feel the new solution make the code much complex. Or do I misunderstand your suggestion? @lukaszlew

Comment (kaixih) [reply paxml/trainer_lib.py]: I think this is actually easier than I thought. I have pushed a new commit to move the updating of the overwrite-with-grads params into the apply_gradients. So, I think this comment can be addressed. (Sorry for the confusion.) @lukaszlew

File: paxml/trainer_lib.py (hunk 3 above)
Comment (kaixih) [reply paxml/trainer_lib.py]: Yes, this is removed.

File: paxml/trainer_lib.py [hunk 5]
```diff
@@ -1130,17 +1130,41 @@ def train_step_single_learner(
...
     )
+
+    # Except for the overwrite_with_gradient params, we use the update_states()
+    # to compute the optimizer states and transformed grads.
+    overwrite_var_mask = \
+        tasks_lib.get_overwrite_with_gradient_var_mask(var_weight_hparams)
+    grads_with_opt = tasks_lib.filter_vars_for_grad_or_opt(
+        grads, overwrite_var_mask
+    )
     transformed_grads, new_opt_states = learner.update_states(
```
Comment (lukaszlew) [inline paxml/trainer_lib.py:1141]: Let's make the update_states and apply_gradient API more uniform. I.e. either both should be aware of OWG variables (and then the learner is fully responsible for OWG). Or none should, and then trainer is fully responsible for OWG.

Comment (kaixih) [reply paxml/trainer_lib.py:1135]: Done.

File: paxml/tasks_lib.py [hunk 6]
```diff
@@ -826,7 +843,7 @@ def get_excluded_var_mask_for_opt(
 ) -> NestedMap:
```
[the remark this answers is not in the document]
Comment (kaixih) [reply paxml/tasks_lib.py:833]: Certainly, as we discussed, regardless of the situation, OWG should not have any connection with the optimizer and optimizer states.

File: paxml/learners.py [hunk 7]
```diff
@@ -416,6 +438,9 @@ def _adjust_var(old_var, transformed_grad, is_learnable, var_wh, var_key):
       if not is_learnable:
         return old_var
 
+      if base_layer.var_overwrite_with_gradient(var_wh):
```
Comment (lukaszlew) [inline paxml/learners.py:441]: We just need to attach documentation somewhere in the repo.
`var_overwrite_with_gradient` function seems like the best place.

**Review by lukaszlew — COMMENTED** [review]
This is great. Just documentation is left. (and some local code "beautification")

File: paxml/learners.py [hunk 8]
```diff
@@ -336,6 +336,21 @@ def update_states(
...
+    # Omit the `overwrite_with_gradient` parameters as they are unnecessary for
+    # the optimizer state update. Gradient overwriting takes place at the very
+    # end of the function to update the transformed gradients.
+    var_is_overwrite_with_grad = jax.tree_util.tree_map(
+        lambda x: base_layer.var_overwrite_with_gradient(x), var_weight_hparams)
+    def _filter_var(old):
+      return jax.tree_map(
+          lambda v, e: py_utils.BpropMaskedNode() if e else v, old,
+          var_is_overwrite_with_grad
+      )
```
[the remark this answers is not in the document]
Comment (kaixih) [reply paxml/learners.py:351]: Done.

File: paxml/learners.py [hunk 9]
```diff
@@ -368,6 +383,13 @@ def _update(updated, original):
           SummaryType.AGGREGATE_SCALAR,
       )
 
+    transformed_grad = jax.tree_map(
```
[the remark this answers is not in the document]
Comment (kaixih) [reply paxml/learners.py:387]: Done

**Review by lukaszlew — APPROVED** [review]
I think this PR is great. 
Let's add documentation in the first PR
https://github.com/google/praxis/pull/29

**Review by zhangqiaorjc — APPROVED** [review]
